mirgecom/filter.py

```diff
@@ -0,0 +1,200 @@
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import grudge.dof_desc as dof_desc
+
+from meshmode.dof_array import DOFArray
+from pytools import keyed_memoize_in
+from pytools.obj_array import obj_array_vectorized_n_args
+
+
+def exponential_mode_response_function(mode, alpha, cutoff, nfilt, filter_order):
+    """Return an exponential filter coefficient for the user-provided *mode*."""
+    return np.exp(-1.0 * alpha * ((mode - cutoff) / nfilt)
+                  ** (2*filter_order))
+
+
+def make_spectral_filter(actx, group, cutoff, mode_response_function):
+    r"""
+    Create a spectral filter with the provided *mode_response_function*.
+
+    This routine returns a filter operator in the modal basis designed to
+    apply the user-provided *mode_response_function* to the spectral modes
+    beyond the user-provided *cutoff*.
+
+    Parameters
+    ----------
+    actx: :class:`meshmode.array_context.ArrayContext`
+        A :class:`meshmode.array_context.ArrayContext` associated with
+        an array of degrees of freedom
+    group: :class:`meshmode.mesh.MeshElementGroup`
+        A :class:`meshmode.mesh.MeshElementGroup` from which the mode ids,
+        element order, and dimension may be retrieved.
+    cutoff: integer
+        Mode cutoff beyond which the filter will be applied, and below which
+        the filter will preserve.
+    mode_response_function:
+        A function that returns a filter weight for for each mode id.
+
+    Returns
+    -------
+    filter: np.ndarray
+        filter operator in the modal basis
+    """
+
+    @keyed_memoize_in(
+        actx, (make_spectral_filter,
+               mode_response_function,
+               "_spectral_filter_matrix"),
+        lambda grp: grp.discretization_key()
+    )
+    def _spectral_filter_matrix(group):
+        mode_ids = group.mode_ids()
+        order = group.order
+
+        nmodes = len(mode_ids)
+        filter_mat = np.identity(nmodes)
+        nfilt = order - cutoff
+        if nfilt <= 0:
+            return filter
+
+        for mode_index, mode_id in enumerate(mode_ids):
+            mode = sum(mode_id)
+            if mode >= cutoff:
+                filter_mat[mode_index, mode_index] = \
+                    mode_response_function(mode,
+                                           cutoff=cutoff,
+                                           nfilt=nfilt)
+
+        return actx.freeze(actx.from_numpy(filter_mat))
+
+    return _spectral_filter_matrix(group)
```

Comment:
This does way more flops than needed. (i.e. no need for a matrix if you're just wanting to scale stuff)

Reply:
@inducer Ah, right. So how about we just build a scaling vector? Then that will simplify the `apply_filter_matrix` method (won't actually do a matvec, just element-wise multiply)

Reply:
Yep, exactly.

Reply:
Done: https://github.com/illinois-ceesd/mirgecom/pull/324/commits/ae92e70c602bcd084e4b034287883de953be0f85